File: src/agent_run/capacity/advice.py
"""Advisory recommendations and a noise-tolerant semantic capacity key.

Advice is always advisory: an explicit owner choice wins regardless of risk.
``advice_key`` intentionally drops high-resolution timestamps so it changes
only on material capacity state, not on every collection tick.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Iterable

from .forecast import RISK_HIGH, RISK_MEDIUM, RISK_UNKNOWN, CapacityForecast
from .history import CapacityKey


_REMAINING_BUCKET_PERCENT = 5.0
_RESET_BUCKET_SECONDS = 300


@dataclass(frozen=True)
class CapacityAdvice:
    key: CapacityKey
    known: bool
    remaining_percent: float | None
    reset_at: float | None
    warmup: bool
    risk: str
    recommendations: tuple[str, ...]


def build_advice(forecasts: Iterable[CapacityForecast]) -> tuple[CapacityAdvice, ...]:
    return tuple(
        CapacityAdvice(
            key=forecast.key,
            known=forecast.known,
            remaining_percent=forecast.remaining_percent,
            reset_at=forecast.reset_at,
            warmup=forecast.warmup,
            risk=forecast.risk,
            recommendations=_recommendations(forecast),
        )
        for forecast in forecasts
    )


def _recommendations(forecast: CapacityForecast) -> tuple[str, ...]:
    label = f"{forecast.key.runtime}/{forecast.key.lane} {forecast.key.window}"
    if forecast.risk == RISK_UNKNOWN:
        return (f"{label} capacity is unknown; treat limits as unverified.",)
    if forecast.risk == RISK_HIGH:
        return (f"{label} is near exhaustion; avoid starting new {forecast.key.lane} work before reset.",)
    if forecast.risk == RISK_MEDIUM:
        return (f"{label} is trending toward exhaustion; pace new requests.",)
    return ()


def _bucketed_remaining(remaining: float | None) -> float | None:
    if remaining is None:
        return None
    return round(remaining / _REMAINING_BUCKET_PERCENT) * _REMAINING_BUCKET_PERCENT


def _bucketed_reset(reset_at: float | None) -> int | None:
    if reset_at is None:
        return None
    return int(reset_at // _RESET_BUCKET_SECONDS)


def _sort_key(advice: CapacityAdvice) -> tuple[str, str, str, str, str]:
    key = advice.key
    return (key.runtime, key.lane, key.window, key.target or "", key.source)


def advice_key(items: Iterable[CapacityAdvice]) -> str:
    ordered = sorted(items, key=_sort_key)
    parts = [
        "|".join(
            (
                key.runtime,
                key.lane,
                key.window,
                key.target or "",
                key.source,
                item.risk,
                "warmup" if item.warmup else "steady",
                str(_bucketed_remaining(item.remaining_percent)),
                str(_bucketed_reset(item.reset_at)),
            )
        )
        for item, key in ((item, item.key) for item in ordered)
    ]
    digest = hashlib.sha256("::".join(parts).encode("utf-8")).hexdigest()
    return digest[:32]

File: crates/agent-run-core/src/capacity/advice.rs
//! Advisory recommendations and a noise-tolerant semantic capacity key.
//!
//! Port of `agent_run.capacity.advice`. Advice is always advisory: an
//! explicit owner choice wins regardless of risk. `advice_key` intentionally
//! drops high-resolution timestamps so it changes only on material capacity
//! state, not on every collection tick.
use super::{Forecast, Key};
use sha2::{Digest, Sha256};

const REMAINING_BUCKET_PERCENT: f64 = 5.0;
const RESET_BUCKET_SECONDS: i64 = 300;

/// One forecast turned into owner-facing advice.
#[derive(Debug, Clone, PartialEq)]
pub struct CapacityAdvice {
    pub key: Key,
    pub known: bool,
    pub remaining_percent: Option<f64>,
    pub reset_at: Option<f64>,
    pub warmup: bool,
    pub risk: String,
    pub recommendations: Vec<String>,
}

pub fn build_advice(forecasts: &[Forecast]) -> Vec<CapacityAdvice> {
    forecasts
        .iter()
        .map(|f| CapacityAdvice {
            key: f.key.clone(),
            known: f.known,
            remaining_percent: f.remaining_percent,
            reset_at: f.reset_at,
            warmup: f.warmup,
            risk: f.risk.clone(),
            recommendations: recommendations(f),
        })
        .collect()
}

pub fn capacity_label(key: &Key) -> String {
    let mut parts = vec![format!("{}/{}", key.runtime, key.lane), key.window.clone()];
    if let Some(target) = &key.target {
        if !target.is_empty() {
            parts.push(format!("target={target}"));
        }
    }
    if !key.source.is_empty() {
        parts.push(format!("source={}", key.source));
    }
    parts.join(" ")
}

fn recommendations(forecast: &Forecast) -> Vec<String> {
    let label = capacity_label(&forecast.key);
    match forecast.risk.as_str() {
        "unknown" => vec![format!(
            "{label} capacity is unknown; treat limits as unverified."
        )],
        "high" => vec![format!(
            "{label} is near exhaustion; avoid starting new {} work before reset.",
            forecast.key.lane
        )],
        "medium" => vec![format!(
            "{label} is trending toward exhaustion; pace new requests."
        )],
        _ => Vec::new(),
    }
}

fn bucketed_remaining(remaining: Option<f64>) -> Option<i64> {
    remaining.map(|r| (r / REMAINING_BUCKET_PERCENT).round() as i64)
}

fn bucketed_reset(reset_at: Option<f64>) -> Option<i64> {
    reset_at.map(|r| (r / RESET_BUCKET_SECONDS as f64).floor() as i64)
}

fn sort_key(advice: &CapacityAdvice) -> (&str, &str, &str, &str, &str) {
    (
        &advice.key.runtime,
        &advice.key.lane,
        &advice.key.window,
        advice.key.target.as_deref().unwrap_or(""),
        &advice.key.source,
    )
}

/// A stable, bounded hash of every advisory item's material state.
///
/// Mirrors Python's `advice_key`: sorts by identity, buckets remaining
/// percent and reset time to absorb collection-tick jitter, and truncates
/// the digest to 32 hex characters.
pub fn advice_key(items: &[CapacityAdvice]) -> String {
    let mut ordered: Vec<&CapacityAdvice> = items.iter().collect();
    ordered.sort_by(|a, b| sort_key(a).cmp(&sort_key(b)));
    let parts: Vec<String> = ordered
        .iter()
        .map(|item| {
            let key = &item.key;
            let remaining = bucketed_remaining(item.remaining_percent)
                .map(|v| v.to_string())
                .unwrap_or_else(|| "None".into());
            let reset = bucketed_reset(item.reset_at)
                .map(|v| v.to_string())
                .unwrap_or_else(|| "None".into());
            [
                key.runtime.as_str(),
                key.lane.as_str(),
                key.window.as_str(),
                key.target.as_deref().unwrap_or(""),
                key.source.as_str(),
                item.risk.as_str(),
                if item.warmup { "warmup" } else { "steady" },
                remaining.as_str(),
                reset.as_str(),
            ]
            .join("|")
        })
        .collect();
    let digest = Sha256::digest(parts.join("::").as_bytes());
    format!("{digest:x}")[..32].to_string()
}
